### PR TITLE
Added adapter for xgboost XGBClassifier - tested with xgboost 2.0.3

### PR DESCRIPTION
**Pull request by LeLoMAu**
I've run black command on python files inside te2rules, that is why format is changed for some

### CODE REVIEW

**Comment by groshanlal** [comment]
Regarding the whitespace changes: 
The repo already has some linters like black, flake8 and mypy setup. 
Please run `pre-commit install` to automatically setup the linters on your local repo.
This will make sure your PR follows the same code formatting standards.

I tried checking out the PR locally and observed some minor formatting errors: 
The following flake8 checks fail:
```
te2rules/adapter.py: error: Function is missing a type annotation for one or more arguments
te2rules/adapter.py: error: Item "None" of "Optional[Match[Any]]" has no attribute "group"
Found 2 errors in 1 file (checked 5 source files)
 ```

Can you please run `pre-commit install` and then push another commit on top after resolving the formatting issues?

File: te2rules/adapter.py [hunk 1]
```diff
@@ -229,3 +233,121 @@ def _convert(self) -> DecisionTree:
...
+    Usage:
+    adapter = XgboostXGBClassifierAdapter(model, feature_names)
+    adapted_model = adapter.random_forest
+    """
+
+    def __init__(self, xgb_model: XGBClassifier, feature_names: List[str]):
+        self.xgb_model = xgb_model
+        self.feature_names = feature_names
+
+        # TODO: n0 and n1 are not accessible from the xgboost model itself, in this way bias term will be 0
```
Comment (groshanlal) [inline te2rules/adapter.py]: It is true that xgboost library doesn't use bias and weights explicitly. `self.bias` can be directly set to 0 without n0, n1. These variables don't get used anywhere.

File: te2rules/adapter.py [hunk 2]
```diff
@@ -229,3 +233,121 @@ def _convert(self) -> DecisionTree:
...
+        self.n0 = 1
+        self.n1 = 1
+        self.bias = np.log(self.n1 / self.n0)
+
+        # Default learning rate 0.3
+        self.weight = self.xgb_model.get_params()["learning_rate"] or 0.3
```
Comment (groshanlal) [inline te2rules/adapter.py]: Since xgboost doesn't use a separate weight variable, `self.weight` can be set to 1. scikit is slightly different in this aspect. It uses learning rate as tree weights, while xgboost does not.

File: te2rules/adapter.py [hunk 3]
```diff
@@ -229,3 +233,121 @@ def _convert(self) -> DecisionTree:
...
+        self.activation = "sigmoid"
+
+        tree_ensemble = self._extract_tree_ensemble()
+        self.tree_ensemble = tree_ensemble
+
+        self.random_forest = self._convert()
+
+    def _extract_tree_ensemble(self) -> np.ndarray:
```
Comment (groshanlal) [inline te2rules/adapter.py]: This part of the code can be moved into the `_convert` method. Currently, the `_convert` method consumes the output of this method and returns it back as a different object.

File: te2rules/adapter.py [hunk 4]
```diff
@@ -229,3 +233,121 @@ def _convert(self) -> DecisionTree:
...
+                # XGBoost stores variable as f0, f1, f2 etc, therefore we extract in the following way the variable index
+                feature_index = re.search("^f(\d+)", node["split"]).group(1)
+                node_name = self.feature_names[int(feature_index)]
+            else:
+                node_name = node["split"]
+
+            tree.node = TreeNode(
+                node_name=node_name,
+                threshold=(
+                    # TODO: For binary variables threshold is 1, we should have 0.5 instead
```
Comment (groshanlal) [inline te2rules/adapter.py]: I just checked why this happens. Apparently, scikit and xgboost treat feature splits differently. In scikit, the left child corresponds to feature <= threshold while in xgboost, the left child corresponds to feature < threshold. Note, how they treat equality differently. 

I have pushed a new change, in which I have extended the TreeNode to support both "<" and "<=". You can rebase your change on the latest commit to use this  feature.

File: te2rules/adapter.py (hunk 3 above)
Comment (LeLoMAu) [reply te2rules/adapter.py]: Hey @groshanlal not clear here, would you just combine `_extract_tree_ensemble` and `_convert` methods into the same method named `_convert`?

Or would you move execution of `_extract_tree_ensemble` into `_convert`?

**Comment by LeLoMAu** [comment]
Hey @groshanlal, about `pre-commit install`:

- I needed to update isort to 5.12.0 because of this issue https://github.com/home-assistant/core/issues/86892
- I needed to add Optional type to some function parameters into explainer as mypy was failing on that

File: te2rules/adapter.py (hunk 3 above)
Comment (groshanlal) [reply te2rules/adapter.py]: Yes, I would combine `_extract_tree_ensemble` and `_convert` methods into the same method named `_convert`

**Comment by groshanlal** [comment]
Looks good to me! Thanks @LeLoMAu.
